File: ios/HybridNitroWebview.swift
//
//  HybridNitroWebview.swift
//  Pods
//
//  Created by Patrick Kabwe on 24/04/2025.
//

import UIKit

class HybridNitroWebview : HybridNitroWebviewSpec {
    // UIView
    lazy var view: UIView = NitroWebView()
    
    private(set) var nitroWebViewImpl: NitroWebViewImpl?
    
    override init() {
        super.init()
    }
    
    // Props
    var onScriptLoaded: (() -> Void)?
    
    var sourceUrl: String = "" {
        didSet {
            if !sourceUrl.isEmpty {
                if nitroWebViewImpl == nil, let webview = view as? NitroWebView {
                    nitroWebViewImpl = NitroWebViewImpl(webview: webview)
                    nitroWebViewImpl?.webview?.delegate = self
                }
                nitroWebViewImpl?.load(with: sourceUrl)
            }
        }
    }
}


extension HybridNitroWebview: NitroWebViewProtocol {
    func didFinishNavigation(success: Bool) {
        if success {
            onScriptLoaded?()
        }
    }
}

File: android/src/main/java/com/nitrowebview/HybridNitroWebview.kt
package com.nitrowebview

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.uimanager.ThemedReactContext
import com.margelo.nitro.nitrowebview.HybridNitroWebviewSpec

@Keep
@DoNotStrip
class HybridNitroWebview(val context: ThemedReactContext): HybridNitroWebviewSpec(), WebViewListener  {
    // View
    // TODO: fix edge-to-edge
    override val view: NitroWebView = NitroWebView(context)
    lateinit var webViewImpl: NitroWebViewImpl

    // Props
    override var sourceUrl: String
        get() = ""
        set(value) {
            webViewImpl = NitroWebViewImpl(view)
            webViewImpl.webView.listener = this
            webViewImpl.load(value)
        }

    override var onScriptLoaded: (() -> Unit)? = null

    override fun onPageFinished(state: OnPageFinishState) {
        if (state == OnPageFinishState.Success){
            onScriptLoaded?.invoke()
        }
    }
}
